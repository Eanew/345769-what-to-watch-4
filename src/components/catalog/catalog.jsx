import React from "react";

import {connect} from "react-redux";
import {ActionCreator as ScreenActionCreator} from "../../reducer/screen/screen.js";
import {ActionCreator as FilmsActionCreator} from "../../reducer/films/films.js";
import {Operation} from "../../reducer/data/data.js";
import Selector from "../../reducer/films/selectors.js";

import pt from "../../prop-types-cover.js";

import Genres from "../genres/genres.jsx";
import Films from "../films/films.jsx";
import ShowMoreButton from "../show-more-button/show-more-button.jsx";

const Catalog = (props) => {
  const {
    genres,
    currentGenre,
    filteredFilms,
    displayedFilms,
    onFilmCardClick,
    onGenreTabClick,
    onShowMoreButtonClick,
  } = props;

  return (
    <section className="catalog">
      <h2 className="catalog__title visually-hidden">Catalog</h2>

      <Genres
        genres={genres}
        currentGenre={currentGenre}
        onGenreTabClick={onGenreTabClick}
      />

      <Films
        films={displayedFilms}
        onFilmCardClick={onFilmCardClick}
      />

      {displayedFilms.length < filteredFilms.length && (
        <ShowMoreButton
          onClick={onShowMoreButtonClick}
        />
      )}

    </section>
  );
};

Catalog.propTypes = {
  genres: pt.genres,
  currentGenre: pt.genre,
  filteredFilms: pt.films,
  displayedFilms: pt.films,
  onGenreTabClick: pt.func,
  onFilmCardClick: pt.func,
  onShowMoreButtonClick: pt.func,
};

const mapStateToProps = (state) => ({
  genres: Selector.getGenresList(state),
  currentGenre: Selector.getCurrentGenre(state),
  filteredFilms: Selector.getFilmsByCurrentGenre(state),
  displayedFilms: Selector.getDisplayedFilms(state),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreTabClick(genre) {
    dispatch(FilmsActionCreator.switchGenre(genre));
  },
  onFilmCardClick(film) {
    dispatch(ScreenActionCreator.setMoviePageScreen(film));
    dispatch(Operation.loadReviews(film.id));
  },
  onShowMoreButtonClick() {
    dispatch(FilmsActionCreator.showMoreFilms());
  },
});

export {Catalog};
export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
