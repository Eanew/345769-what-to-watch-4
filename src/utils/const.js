export const Screen = {
  MAIN: `main`,
  DETAILS: `details`,
};

export const APPROVED_GENRES = [
  `Comedy`,
  `Crime`,
  `Documentary`,
  `Drama`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thriller`];

export const Rating = {
  'Awesome': 10,
  'Very good': 8,
  'Good': 5,
  'Normal': 3,
  'Bad': 0,
};
