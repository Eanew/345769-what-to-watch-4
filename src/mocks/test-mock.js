import {toKebabCase} from "../utils/common.js";
import {APPROVED_GENRES} from "../utils/const.js";

const filmsTitles = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`];

const usersNames = [
  `Viktor Savinov`,
  `Alexandra Pushina`,
  `Julia Kim`,
  `Anton Timoshenkov`];

export const promo = {
  filmTitle: filmsTitles[0],
  genre: APPROVED_GENRES[0],
  release: 2014,
};

export const film = {
  id: filmsTitles[0] + 1,
  filmTitle: filmsTitles[0],
  release: 2011,
  runtime: 143,
  genre: APPROVED_GENRES[0],

  rating: {
    value: 6.3,
    votesCount: 2134,
  },

  image: {
    preview: `img/${toKebabCase(filmsTitles[0])}.jpg`,
    background: `img/bg-${toKebabCase(filmsTitles[0])}.jpg`,
    poster: `img/${toKebabCase(filmsTitles[0])}-poster.jpg`,
  },

  movie: {
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    full: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  },

  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,

  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  isFavorite: false,
};

export const films = filmsTitles.map((title, i) => ({
  id: title + i,
  filmTitle: title,
  release: 2011 + i,
  runtime: 30 + i * 60,
  genre: APPROVED_GENRES[i],

  rating: {
    value: 3.4 + i,
    votesCount: 153 + i * 421,
  },

  image: {
    preview: `img/${toKebabCase(title)}.jpg`,
    background: `img/bg-${toKebabCase(title)}.jpg`,
    poster: `img/${toKebabCase(title)}-poster.jpg`,
  },

  movie: {
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    full: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  },

  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,

  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  isFavorite: i % 2,
}));

export const review = {
  id: `comment1`,
  userName: usersNames[0],
  rating: 6.3,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
};

export const reviews = usersNames.map((name, i) => ({
  id: `comment${i + 1}`,
  userName: name,
  rating: 3.2 + i,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
}));
