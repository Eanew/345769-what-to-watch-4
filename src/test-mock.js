import {toKebabCase} from "./utils/common.js";
import {Genre, Screen, MovieTab} from "./utils/const.js";
import NameSpace from "./reducer/name-space.js";

const filmsTitles = [
  `Fantastic Beasts: The Crimes of Grindelwald`,
  `Bohemian Rhapsody`,
  `Macbeth`];

const usersNames = [
  `Viktor Savinov`,
  `Alexandra Pushina`,
  `Julia Kim`,
  `Anton Timoshenkov`];

export const userInfo = {
  isAuthorized: true,
  id: 1,
  name: usersNames[2],
  email: `julia.kim@gmail.com`,
  avatar: `img/avatar.jpg`,
};

export const film = {
  id: filmsTitles[0] + 1,
  filmTitle: filmsTitles[0],
  release: 2011,
  runtime: 143,
  genre: Genre.SCI_FI,

  rating: {
    value: 6.3,
    votesCount: 2134,
  },

  image: {
    preview: `img/${toKebabCase(filmsTitles[0])}.jpg`,
    poster: `img/${toKebabCase(filmsTitles[0])}-poster.jpg`,
    background: `img/bg-${toKebabCase(filmsTitles[0])}.jpg`,
    backgroundColor: `#ffffff`,
  },

  movie: {
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    full: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  },

  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,

  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  isFavorite: false,
};

export const films = filmsTitles.map((title, i) => ({
  id: title + i,
  filmTitle: title,
  release: 2011 + i,
  runtime: 30 + i * 60,
  genre: Object.values(Genre)[i],

  rating: {
    value: 3.4 + i,
    votesCount: 153 + i * 421,
  },

  image: {
    preview: `img/${toKebabCase(title)}.jpg`,
    poster: `img/${toKebabCase(title)}-poster.jpg`,
    background: `img/bg-${toKebabCase(title)}.jpg`,
    backgroundColor: `#ffffff`,
  },

  movie: {
    preview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    full: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  },

  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,

  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  isFavorite: (i % 2) ? true : false,
}));

export const review = {
  id: `comment1`,
  userId: 1,
  userName: usersNames[0],
  rating: 6.3,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
};

export const reviews = usersNames.map((name, i) => ({
  id: i + 1,
  userId: (i + 1) * 3,
  userName: name,
  rating: 3.2 + i,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
}));

export const mockStore = {
  [NameSpace.DATA]: {
    promo: films[0],
    films,
    favorites: films.filter((movie) => movie.isFavorite),
    reviews,
  },

  [NameSpace.USER]: {
    userInfo,
  },

  [NameSpace.FILMS]: {
    genre: Genre.ALL,
    displayedFilmsCount: 8,
  },

  [NameSpace.SCREEN]: {
    screen: Screen.MAIN,
    lastScreen: Screen.MAIN,
    movieTab: MovieTab.OVERVIEW,
  },
};
