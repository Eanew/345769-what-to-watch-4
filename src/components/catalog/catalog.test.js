import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";

const store = configureStore([]);

import {films, mockStore} from "../../test-mock.js";

import {Genre} from "../../utils/const.js";

import {Catalog} from "./catalog.jsx";

it(`Should CatalogComponent render correctly`, () => {
  const tree = renderer.create(
      <Provider store={store(mockStore)}>
        <Catalog
          genres={Object.values(Genre).sort().slice(0, 10)}
          currentGenre={Genre.ALL}
          filteredFilms={films}
          displayedFilms={films.slice(0, 8)}
          onGenreTabClick={() => {}}
          onFilmCardClick={() => {}}
          onShowMoreButtonClick={() => {}}
        />
      </Provider>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
