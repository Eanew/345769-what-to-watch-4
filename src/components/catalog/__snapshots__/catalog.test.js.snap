// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should CatalogComponent render correctly 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Actions
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Adventures
      </a>
    </li>
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Comedies
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Documentary
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Dramas
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Fantasy
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Horror
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Kids & Family
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          controls={false}
          height="175"
          loop={true}
          muted={true}
          poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
          preload="none"
          width="280"
        >
          <source
            src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href=""
        >
          Fantastic Beasts: The Crimes of Grindelwald
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          controls={false}
          height="175"
          loop={true}
          muted={true}
          poster="img/bohemian-rhapsody.jpg"
          preload="none"
          width="280"
        >
          <source
            src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href=""
        >
          Bohemian Rhapsody
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          controls={false}
          height="175"
          loop={true}
          muted={true}
          poster="img/macbeth.jpg"
          preload="none"
          width="280"
        >
          <source
            src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            type="video/mp4"
          />
        </video>
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href=""
        >
          Macbeth
        </a>
      </h3>
    </article>
  </div>
</section>
`;
